### PR TITLE
SLING-11882 shade guava library to avoid classpath problems in unit tests

### PR DESCRIPTION
**Pull request by stefanseifert**
https://issues.apache.org/jira/browse/SLING-11882

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -172,6 +172,33 @@
...
+                    <artifactSet>
+                        <includes>
+                            <include>com.google.guava:*</include>
+                        </includes>
+                    </artifactSet>
+                    <createSourcesJar>true</createSourcesJar>
+                    <relocations>
+                        <relocation>
+                            <pattern>com.google.common</pattern>
+                            <shadedPattern>sling-xss.com.google.common</shadedPattern>
```
Comment (rombert) [inline pom.xml]: Is this shadedPattern valid? Java package names cannot contain the `-` / dash character.

( https://docs.oracle.com/javase/specs/jls/se16/html/jls-3.html#jls-Identifier , https://docs.oracle.com/javase/specs/jls/se16/html/jls-7.html#jls-7.4.1 )

**Review by rombert — COMMENTED** [review]
LGTM, but have a question regarding the shaded pattern.

File: pom.xml (hunk 1 above)
Comment (stefanseifert) [reply pom.xml]: i stumbled over this as well when taking over the configuration from sling models impl bundle - it seems to work fine there, and i deployed this bundle in a local instance and it seems to work there as well. but we can also easily remove the dash here.

Comment (stefanseifert) [reply pom.xml]: removed the dash

**Review by rombert — APPROVED** [review]

File: pom.xml (hunk 1 above)
Comment (rombert) [reply pom.xml]: Thanks for the change. I think this is a restriction of Java at the source level, but it seems to be legal bytecode. Before your change, the compiled java classes reference the 'sling-xss' package, see below snippet

```
$ javap -verbose -cp target/org.apache.sling.xss-2.3.9-SNAPSHOT.jar org.apache.sling.xss.impl.HtmlSanitizer
(...)
        47: invokespecial #86                 // Method org/owasp/html/DynamicAttributesSanitizerPolicy."<init>":(Lorg/owasp/html/HtmlStreamEventReceiver;Lsling-xss/com/google/common/collect/ImmutableMap;Lsling-xss/com/google/common/collect/ImmutableSet;Ljava/util/Map;Ljava/util/List;)V
(...)
````

More a curiosity than something to spend more time about.

**Comment by enapps-enorman** [comment]
Is it really still necessary to embed another copy of guava in this bundle?  If the reason for embedding guava was that it required a newer version than was required by oak, then has that not been resolved with OAK-9989?  Just referencing the oak-shaded-guava instead of shading and embedding another private copy again would make this bundle much smaller.

**Comment by stefanseifert** [comment]
in this case guava is a dependency of the also embedded https://github.com/OWASP/java-html-sanitizer, so we need it here and even in a specific version. maybe we can get rid of it when if https://github.com/OWASP/java-html-sanitizer/pull/272 gets resolved.

**Comment by enapps-enorman** [comment]
> in this case guava is a dependency of the also embedded https://github.com/OWASP/java-html-sanitizer, so we need it here and even in a specific version. maybe we can get rid of it when if [OWASP/java-html-sanitizer#272](https://github.com/OWASP/java-html-sanitizer/pull/272) gets resolved.

It would definitely be great to see people stop using guava when it is not necessary.  But I'm not sure your conclusion is quite right.  

I assume the guava dependency is more of a "minimum" version (30.1-jre?) rather than a specific version. The 32.0.1-jre version in oak-shaded-guava should be compatible?  In that case, then if you configure the maven-shade-plugin to use the same re-written package name as the oak-shaded-guava uses, then it would rewrite the third-party binary bytecode to be compatible.  At that point, there should be no need to embed any of the com.google.common.* as a private package here.  The runtime could resolve those from the oak-shaded-guava bundle instead and the xss bundle would go from ~4MB to ~2MB without losing any functionality.

**Comment by stefanseifert** [comment]
best thing of course would be to get rid of guava completely, but this is out of our hands as long as we want to use java-html-sanitizer for xss.
imho using oak-shaded-guava is not a good idea - as the name suggest this should only be used by oak, and may be updated (or removed) by the oak team at any time. and guava does not play nicely with OSGi package versions (https://github.com/google/guava/issues/1682), increasing all packages with major version updates even if not needed.

(this discussion should've been placed in [SLING-7231](https://issues.apache.org/jira/browse/SLING-7231), which was released long ago. in this PR, we're just shading the artifact, it was embedded before already.)

**Comment by rombert** [comment]
Speaking as the main reviewer for SLING-7231, I fully agree that we should stay away from Guava. That change was made when it became clear that AntiSamy was not the right tool for us anymore. It required us to play all sorts of weird games with embedding XML libraries and fiddling with classloaders. Some of those embedded libraries are both EOL and affected by critical CVEs.

We switched to the OWASP Java HTML Sanitizer which looks better but (IMO) also does not see that much activity, so I think the Guava dependency is here to stay in the medium run, unfortunately.

While, as a rule, we should try and import dependant bundles, we are embedding the Sanitizer (and playing tricks with the with it ( #32 ) ) so I think we should also keep the Guava dependency embedded in the bundle.

oak-shaded-guava sounds like Oak trying allow other libraries to use which Guava library they want, which is great. In that direction, I think that Sling bundles should not require Guava at all and allow downstream applications do deploy whichever version of Guava they need.

Looking the latest Sling Starter, I see only the Oak bundles and our own Oak-Server import Guava pacakges

image

Once Oak is "Guava-free", it might be a good idea for us to use it from Oak-Server as well and then drop our copy of Guava from the Starter.

------

On a final note, with the new library the XSS bundle is still very large, but down by a more than 40%

```
$ curl --silent --head https://repo1.maven.org/maven2/org/apache/sling/org.apache.sling.xss/2.2.20/org.apache.sling.xss-2.2.20.jar | grep '^content-length'
content-length: 6831549
curl --silent --head https://repo1.maven.org/maven2/org/apache/sling/org.apache.sling.xss/2.3.8/org.apache.sling.xss-2.3.8.jar | grep '^content-length'
content-length: 4008615
```

Hopefully once upstream drops Guava we can reduce the bundle size even more.

**Pull request merged**